### PR TITLE
Use optimized osd store for new clusters

### PR DESCRIPTION
**Pull request by sp98**
If `ocs.openshift.io/clusterIsDisasterRecoveryTarget: true` annotation is available in the StorageCluster, then create `cephCluster` using `bluestore-rdr` backend store. 

If storageCluster is some how updated to remove this annotation, then ensure that
`bluestore-rdr` store type is not overridden

This PR also reverts the changes that exposed the OSD store in the storageCluster spec, that is #2116.

### CODE REVIEW

File: api/v1/storagecluster_types.go [hunk 1]
```diff
@@ -686,3 +690,14 @@ func (r *StorageCluster) NewToolsDeployment(tolerations []corev1.Toleration) *ap
...
+// OptimizeDisasterRecovery returns true if any RDR optimizations are required
+func (r *StorageCluster) OptimizeDisasterRecovery() bool {
+	if annotation, found := r.GetAnnotations()[DisasterRecoveryOptimizations]; found {
+		if annotation == "true" {
+			return true
+		}
+	}
+
+	return false
+}
```
Comment (umangachapagain) [inline api/v1/storagecluster_types.go:703]: Let's move this to `controllers/storagecluster/cephcluster.go` or to the utility package `controllers/util/`

**Review by umangachapagain — CHANGES_REQUESTED** [review]

File: api/v1/storagecluster_types.go (hunk 1 above)
Comment (sp98) [reply api/v1/storagecluster_types.go:703]: moved it to `controllers/storagecluster/cephcluster.go`

File: api/v1/storagecluster_types.go [hunk 2]
```diff
@@ -29,6 +29,10 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+const (
+	DisasterRecoveryOptimizations = "ocs.openshift.io/optimizeDisasterRecovery"
+)
```
[the remark this answers is not in the document]
Comment (sp98) [reply api/v1/storagecluster_types.go:34]: moved it to `controllers/storagecluster/cephcluster.go`. Didn't find a right place within utils package

File: controllers/storagecluster/cephcluster.go [hunk 3]
```diff
@@ -1255,3 +1259,12 @@ func getIPFamilyConfig(c client.Client) (rookCephv1.IPFamilyType, bool, error) {
 
 	return rookCephv1.IPv4, false, nil
 }
+
+func getOSDStoreConfig(sc *ocsv1.StorageCluster) rookCephv1.OSDStore {
+	osdStore := rookCephv1.OSDStore{}
+	if !sc.Spec.ExternalStorage.Enable && sc.OptimizeDisasterRecovery() {
+		osdStore.Type = string(rookCephv1.StoreTypeBlueStoreRDR)
+	}
```
[the remark this answers is not in the document]
Comment (sp98) [reply controllers/storagecluster/cephcluster.go]: what might cause the deletion of the annotation, other than a user manually deleting it?

[the remark this answers is not in the document]
Comment (sp98) [reply controllers/storagecluster/cephcluster.go]: only way I could think of preventing it is to add some validations before the cephCluster is updated and prevent changing osd storetype if `bluestoreRDR` is already applied. But looks hackish

[the remark this answers is not in the document]
Comment (umangachapagain) [reply controllers/storagecluster/cephcluster.go]: > what might cause the deletion of the annotation, other than a user manually deleting it?
>
Users manually deleting or other automation overriding it.

> only way I could think of preventing it is to add some validations before the cephCluster updated operation and prevent changing osd storetype if bluestoreRDR is already applied. But looks hackish

Yes, that's pretty much what I thought.

```
if osdStore is not bluestoreRDR:
    osdStore = getOSDStoreConfig(sc)
else
   continue
```
This allows to skip the function call when it's not needed, but leaks implementation of OSDStore to calling function.

```
func  getOSDStoreConfig(...) osdStoreConfig{
 if sc.spec.osdStore == blueStoreRDR{
  return sc.spec.osdStore
 }
 return osdStoreConfig{bluestoreRDR}
}
```
Adds an extra function call for every reconcile, but allows encapsulating OSDStoreConfig implementation. It also allows more changes later within the function.

File: controllers/storagecluster/cephcluster.go [hunk 4]
```diff
@@ -405,6 +410,9 @@ func newCephCluster(sc *ocsv1.StorageCluster, cephImage string, nodeCount int, s
 		MaxLogSize:  &maxLogSize,
 	}
 
+	osdStore := getOSDStoreConfig(sc)
+	reqLogger.Info("osd store settings", osdStore)
```
Comment (travisn) [inline controllers/storagecluster/cephcluster.go:414]: For the non-rdr case, will this setting be blank? Perhaps only log this if not blank?

File: controllers/storagecluster/cephcluster.go [hunk 5]
```diff
@@ -1256,3 +1264,31 @@ func getIPFamilyConfig(c client.Client) (rookCephv1.IPFamilyType, bool, error) {
...
+	if annotation, found := sc.GetAnnotations()[DROptimizationAnnotation]; found {
+		if annotation == "true" {
+			return true
+		}
+	}
+
+	return false
+}
+
+func updateRDROptimizations(newOSDStore, existingOSDStore rookCephv1.OSDStore) rookCephv1.OSDStore {
```
Comment (travisn) [inline controllers/storagecluster/cephcluster.go:1288]: How about this method name?
```suggestion
func determineOSDStore(newOSDStore, existingOSDStore rookCephv1.OSDStore) rookCephv1.OSDStore {
```

**Review by travisn — CHANGES_REQUESTED** [review]

File: controllers/storagecluster/cephcluster.go (hunk 4 above)
Comment (sp98) [reply controllers/storagecluster/cephcluster.go]: yes. Logging only if not blank.

**Review by travisn — COMMENTED** [review]
Remind me, why is it that we're using an annotation on the StorageCluster CR instead of a setting on the same CR?

**Comment by sp98** [comment]
> Remind me, why is it that we're using an annotation on the StorageCluster CR instead of a setting on the same CR?

Earlier implementation used the entire osd store spec in the `storageCluster` spec. So user needed to provide `osd-store` type in the spec. 

Current changes hide these implementation details from the spec. Since this is only a one time operation, user can't really go back after applying the new osd store, we decided to keep it an annotation than setting on the CR spec itself.

**Comment by umangachapagain** [comment]
> Remind me, why is it that we're using an annotation on the StorageCluster CR instead of a setting on the same CR?
>
@travisn What OSD store to use is an implementation detail. We do not expect users to know what it means and we do not want them to manage it. Also, once OSD store is set to "bluestore-rdr", we do not want them to revert it. Since ocs-operator will internally handle this configuration, and users have no real control over it, it didn't seem reasonable as a StorageCluster spec. When we support brownfield, this spec will be even more useless and a maintenance overhead.

I suggested making this an annotation as a stopgap, since, in greenfield deployments, we needed a way to know if ODF needs to be configured for DR. And, this is a one time operation, as once cluster is deployed, we do not support changing configurations.
When we have better automation in place including brownfield support, we can just drop this annotation or ignore it without any additional overhead, which wouldn't be possible with a spec change.

Does it make sense? Or, am I overthinking this?

**Comment by travisn** [comment]
> > Remind me, why is it that we're using an annotation on the StorageCluster CR instead of a setting on the same CR?
> 
> @travisn What OSD store to use is an implementation detail. We do not expect users to know what it means and we do not want them to manage it. Also, once OSD store is set to "bluestore-rdr", we do not want them to revert it. Since ocs-operator will internally handle this configuration, and users have no real control over it, it didn't seem reasonable as a StorageCluster spec. When we support brownfield, this spec will be even more useless and a maintenance overhead.
> 
> I suggested making this an annotation as a stopgap, since, in greenfield deployments, we needed a way to know if ODF needs to be configured for DR. And, this is a one time operation, as once cluster is deployed, we do not support changing configurations. When we have better automation in place including brownfield support, we can just drop this annotation or ignore it without any additional overhead, which wouldn't be possible with a spec change.
> 
> Does it make sense? Or, am I overthinking this?

Yes, sounds good to use an annotation for such a short-term setting. Thanks for the background.

**Review by umangachapagain — APPROVED** [review]

**Pull request merged**